File: backend/src/controllers/menuItemController.ts
import { type Context } from "hono";
import { BaseController } from "#/mvc/baseController";
import {
  route,
  httpGet,
  httpQuery,
  httpPost,
  httpPut,
  httpDelete,
  fromBody,
  fromRoute,
  fromQuery,
  producesResponseType,
  authorized,
} from "#/mvc";
import type { IMenuItemService } from "#/services/features/menuItemService";
import type { IMenuItemApi } from "@hc-management/shared/api";
import {
  MenuItemListRequestDto,
  MenuItemUpsertRequestDto,
  type MenuItemBasicResponseDto,
  type MenuItemDetailResponseDto
} from "@hc-management/shared/dtos";

@route("/api/menu-categories")
@authorized
export class MenuItemController extends BaseController implements IMenuItemApi {
  private readonly menuItemService: IMenuItemService;

  public constructor(httpContext: Context) {
    super(httpContext);
    this.menuItemService = this.serviceProvider.getRequiredService("menuItemService");
  }

  @httpQuery("/")
  @fromQuery(0, MenuItemListRequestDto)
  @producesResponseType<MenuItemBasicResponseDto[]>(200)
  public async getListAsync(requestDto: MenuItemListRequestDto): Promise<MenuItemBasicResponseDto[]> {
    return this.ok(await this.menuItemService.getListAsync(requestDto));
  }

  @httpGet("/:id{[0-9]+}")
  @fromRoute(0, "id", (v) => v.toNumber())
  @producesResponseType<MenuItemDetailResponseDto>(200)
  @producesResponseType(401)
  @producesResponseType(403)
  @producesResponseType(404)
  public async getDetailAsync(id: number): Promise<MenuItemDetailResponseDto> {
    return this.ok(await this.menuItemService.getDetailAsync(id));
  }

  @httpPost("/")
  @fromBody(0, MenuItemUpsertRequestDto)
  @producesResponseType<number>(201)
  @producesResponseType(400)
  @producesResponseType(401)
  @producesResponseType(403)
  @producesResponseType(422)
  public async createAsync(requestDto: MenuItemUpsertRequestDto): Promise<number> {
    const createdId = await this.menuItemService.createAsync(requestDto);
    return this.created(`/api/menu-category/${createdId}`, createdId);
  }

  @httpPut("/:id{[0-9]+}")
  @fromRoute(0, "id", (v) => v.toNumber())
  @fromBody(1, MenuItemUpsertRequestDto)
  @producesResponseType(200)
  @producesResponseType(400)
  @producesResponseType(401)
  @producesResponseType(403)
  @producesResponseType(404)
  @producesResponseType(422)
  public async updateAsync(id: number, requestDto: MenuItemUpsertRequestDto): Promise<void> {
    await this.menuItemService.updateAsync(id, requestDto);
    return this.ok();
  }

  @httpDelete("/:id{[0-9]+}")
  @fromRoute(0, "id", (v) => v.toNumber())
  @fromBody(1, MenuItemUpsertRequestDto)
  @producesResponseType(200)
  @producesResponseType(401)
  @producesResponseType(403)
  @producesResponseType(404)
  @producesResponseType(422)
  public async deleteAsync(id: number): Promise<void> {
    await this.menuItemService.deleteAsync(id);
    return this.ok();
  }
}

File: backend/Api/Controllers/MenuItemController.cs
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using HCManagement.Core.Features.MenuItems;

namespace HCManagement.Api.Controllers;

[ApiController]
[Route("api/menu-items")]
[Authorize]
public class MenuItemController : ControllerBase
{
    #region Fields
    private readonly IMenuItemService _service;
    #endregion

    #region Constructors
    public MenuItemController(IMenuItemService service)
    {
        _service = service;
    }
    #endregion

    #region Methods
    [HttpPost("list")]
    [ProducesResponseType<List<MenuItemBasicResponseDto>>(StatusCodes.Status200OK)]
    [ProducesResponseType(StatusCodes.Status400BadRequest)]
    [ProducesResponseType(StatusCodes.Status401Unauthorized)]
    public async Task<IActionResult> GetListAsync([FromBody] MenuItemListRequestDto requestDto)
    {
        return Ok(await _service.GetListAsync(requestDto));
    }

    [HttpGet("{id:int}")]
    [ProducesResponseType<MenuItemDetailResponseDto>(StatusCodes.Status200OK)]
    [ProducesResponseType(StatusCodes.Status400BadRequest)]
    [ProducesResponseType(StatusCodes.Status401Unauthorized)]
    [ProducesResponseType(StatusCodes.Status404NotFound)]
    public async Task<IActionResult> GetDetailAsync([FromRoute] int id)
    {
        return Ok(await _service.GetDetailAsync(id));
    }

    [HttpPost]
    [ProducesResponseType<int>(StatusCodes.Status201Created)]
    [ProducesResponseType(StatusCodes.Status400BadRequest)]
    [ProducesResponseType(StatusCodes.Status401Unauthorized)]
    [ProducesResponseType(StatusCodes.Status409Conflict)]
    [ProducesResponseType(StatusCodes.Status422UnprocessableEntity)]
    public async Task<IActionResult> CreateAsync([FromBody] MenuItemUpsertRequestDto requestDto)
    {
        int id = await _service.CreateAsync(requestDto);
        return CreatedAtAction(nameof(GetDetailAsync), new { id }, id);
    }

    [HttpPut("{id:int}")]
    [ProducesResponseType(StatusCodes.Status200OK)]
    [ProducesResponseType(StatusCodes.Status400BadRequest)]
    [ProducesResponseType(StatusCodes.Status401Unauthorized)]
    [ProducesResponseType(StatusCodes.Status409Conflict)]
    [ProducesResponseType(StatusCodes.Status422UnprocessableEntity)]
    public async Task<IActionResult> UpdateAsync([FromRoute] int id, [FromBody] MenuItemUpsertRequestDto requestDto)
    {
        await _service.UpdateAsync(id, requestDto);
        return Ok();
    }

    [HttpDelete("{id:int}")]
    [ProducesResponseType(StatusCodes.Status200OK)]
    [ProducesResponseType(StatusCodes.Status401Unauthorized)]
    [ProducesResponseType(StatusCodes.Status404NotFound)]
    [ProducesResponseType(StatusCodes.Status409Conflict)]
    [ProducesResponseType(StatusCodes.Status422UnprocessableEntity)]
    public async Task<IActionResult> DeleteAsync([FromRoute] int id)
    {
        await _service.DeleteAsync(id);
        return Ok();
    }

    [HttpGet("default-list-parameters")]
    [ProducesResponseType<MenuItemListRequestDto>(StatusCodes.Status200OK)]
    public async Task<IActionResult> GetDefaultListParamtersAsync()
    {
        await Task.CompletedTask;
        return Ok(new MenuItemListRequestDto());
    }
    #endregion
}
 